vm/compiler/src/ledger/helpers/mod.rs

```diff
@@ -14,281 +14,340 @@
 // You should have received a copy of the GNU General Public License
 // along with the snarkVM library. If not, see <https://www.gnu.org/licenses/>.
 
-// TODO (raychu86): Transition out of floating point arithmetic to something more precise.
-
-// TODO (raychu86): Handle downcasting.
+use super::*;
 
 #[allow(unused)]
+///
 /// Calculate the staking reward, given the starting supply and anchor time.
-pub(crate) fn staking_reward<const STARTING_SUPPLY: u64, const ANCHOR_TIME: u64>() -> u64 {
-    // The staking percentage at genesis.
-    const STAKING_PERCENTAGE: f64 = 0.025f64; // 2.5%
-
-    let block_height_around_year_1 = estimated_block_height(ANCHOR_TIME, 1);
-
-    let reward = (STARTING_SUPPLY as f64 * STAKING_PERCENTAGE) / block_height_around_year_1 as f64;
-
-    reward.floor() as u64
+///     R_staking = floor((0.025 * S) / H_Y1)
+///     S = Starting supply.
+///     H_Y1 = Expected block height at year 1.
+///
+pub(crate) fn staking_reward<const STARTING_SUPPLY: u64, const ANCHOR_TIME: i64>() -> Result<u64> {
+    // Calculate the estimated block height at year 1.
+    let block_height_around_year_1 = estimated_block_height(ANCHOR_TIME as u64, 1);
+
+    Ok(u64::try_from(STARTING_SUPPLY as u128 * 25 / 1000 / block_height_around_year_1 as u128)?)
 }
 
-/// Calculate the proving reward for a given block.
-pub(crate) fn proving_reward<const STARTING_SUPPLY: u64, const ANCHOR_TIMESTAMP: u64, const ANCHOR_TIME: u64>(
-    num_validators: u64,
-    timestamp: u64,
+///
+/// Calculate the coinbase reward for a given block.
+///     R_coinbase = max(0, H_Y10 - H) * R_anchor * 2^((D - B) / N).
```

Comment:
This comment seems conflict with the implemention and comment below. I think the last factor should be 
`2^((B - D) / N)` 
or 
`2^(-1 * ((B - D) / N))` 
here.

Reply:
Thanks for catching this!